tests/test_trusted_metadata_set.py

```diff
@@ -1,122 +1,330 @@
+        self.assertTrue(len(self.trusted_set), 6)
+
+    def test_out_of_order_ops(self):
         # Update timestamp before root is finished
         with self.assertRaises(RuntimeError):
-            trusted_set.update_timestamp(data["timestamp"])
+            self.trusted_set.update_timestamp(self.metadata["timestamp"])
 
-        trusted_set.root_update_finished()
+        self.trusted_set.root_update_finished()
         with self.assertRaises(RuntimeError):
-            trusted_set.root_update_finished()
+            self.trusted_set.root_update_finished()
+
+        # Update root after a previous successful root update
+        with self.assertRaises(RuntimeError):
+            self.trusted_set.update_root(self.metadata["root"])
 
         # Update snapshot before timestamp
         with self.assertRaises(RuntimeError):
-            trusted_set.update_snapshot(data["snapshot"])
+            self.trusted_set.update_snapshot(self.metadata["snapshot"])
 
-        trusted_set.update_timestamp(data["timestamp"])
+        self.trusted_set.update_timestamp(self.metadata["timestamp"])
 
         # Update targets before snapshot
         with self.assertRaises(RuntimeError):
-            trusted_set.update_targets(data["targets"])
+            self.trusted_set.update_targets(self.metadata["targets"])
 
-        trusted_set.update_snapshot(data["snapshot"])
+        self.trusted_set.update_snapshot(self.metadata["snapshot"])
 
-        #update timestamp after snapshot
+        # update timestamp after snapshot
         with self.assertRaises(RuntimeError):
-            trusted_set.update_timestamp(data["timestamp"])
+            self.trusted_set.update_timestamp(self.metadata["timestamp"])
 
         # Update delegated targets before targets
         with self.assertRaises(RuntimeError):
-            trusted_set.update_delegated_targets(data["role1"], "role1", "targets")
+            self.trusted_set.update_delegated_targets(
+                self.metadata["role1"], "role1", "targets"
+            )
 
-        trusted_set.update_targets(data["targets"])
-        trusted_set.update_delegated_targets(data["role1"], "role1", "targets")
+        self.trusted_set.update_targets(self.metadata["targets"])
 
-    def test_update_with_invalid_json(self):
-        repo_dir = os.path.join(os.getcwd(), 'repository_data', 'repository', 'metadata')
-        data={}
-        for md in ["root", "timestamp", "snapshot", "targets", "role1"]:
-            with open(os.path.join(repo_dir, f"{md}.json"), "rb") as f:
-                data[md] = f.read()
+        # Update snapshot after sucessful targets update
+        with self.assertRaises(RuntimeError):
+            self.trusted_set.update_snapshot(self.metadata["snapshot"])
+
+        self.trusted_set.update_delegated_targets(
+            self.metadata["role1"], "role1", "targets"
+        )
+
+        self.trusted_set.update_targets(self.metadata["targets"])
```

Comment:
I think this is just an unneeded call...

The call succeeds (even though delegated target role1 has already been loaded) against what's said in the TrustedMetadataSet docs _"Metadata is updatable only if metadata it depends on is loaded"_. We ended up with a version that does not explicitly check that I think because the meta info in snapshot should ensure that the targets metadata is consistent even if a delegating metadata gets "reloaded". That said, there is now a possibility of a repository doing weird things like publishing a new "targets.json" with same version but modified delegations... maybe it's not something we need to care about, but at least the TrustedMetadataSet docs should be updated  (not in this PR)

Reply:
actually the quote from the docs is for top-level metadata so it is technically correct... Might make sense to mention that delegating targets consistency comes from snapshots meta information.

Reply:
> That said, there is now a possibility of a repository doing weird things like publishing a new "targets.json" with same version but modified delegations... maybe it's not something we need to care about, but at least the TrustedMetadataSet docs should be updated (not in this PR)

I guess we don't care about that because if the file is signed from the keys trusted by root.json then everything is fine right? 

We haven't documented which roles are used for verification when calling all other `update_*` functions.
Why do we want to add this specifically here?